src/eckit/geo/Projection.h

```diff
@@ -0,0 +1,122 @@
+/*
+ * (C) Copyright 1996- ECMWF.
+ *
+ * This software is licensed under the terms of the Apache Licence Version 2.0
+ * which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * In applying this licence, ECMWF does not waive the privileges and immunities
+ * granted to it by virtue of its status as an intergovernmental organisation nor
+ * does it submit to any jurisdiction.
+ */
+
+
+#pragma once
+
+#include <string>
+
+#include "eckit/geo/Point.h"
+#include "eckit/memory/Builder.h"
+#include "eckit/memory/Factory.h"
+
+
+namespace eckit::geo {
+class Spec;
+namespace spec {
+class Custom;
+}
+}  // namespace eckit::geo
+
+
+namespace eckit::geo {
+
+
+class Projection {
+public:
+    // -- Types
+
+    using builder_t = BuilderT1<Projection>;
+    using ARG1      = const Spec&;
+
+    // -- Exceptions
+    // None
+
+    // -- Constructors
+
+    Projection() noexcept         = default;
+    Projection(const Projection&) = default;
+    Projection(Projection&&)      = default;
+
+    // -- Destructor
+
+    virtual ~Projection() = default;
+
+    // -- Convertors
+    // None
+
+    // -- Operators
+
+    Projection& operator=(const Projection&) = default;
+    Projection& operator=(Projection&&)      = default;
+
+    // -- Methods
+
+    static std::string className() { return "projection"; }
+
+    virtual Point fwd(const Point&) const = 0;
+    virtual Point inv(const Point&) const = 0;
+
+    virtual Spec* spec() const = 0;
```

Comment:
Same comment: `[[nodiscard]]` or better return `std::unique_ptr<Spec>`.

Reply:
The semantics I use, and that was learned from experience on both the mars-client and pgen stack, is that if a function/method returns a pointer, it is up to the client code to catch it and manage it's life cycle. If the object is to be "used" but not "managed", then a reference suffices. It is very deliberate and throws a warning to any developer -- that they should care.

The use is typically from factories ("build") or type generators ("make"), and spec is the other case: it means that it generates an object of this type. I did consider std::unique_ptr would be useful with the spec generation, because it's the exception of the above, but decided to use the same pattern as a mix would be more confusing than clarifying.

But I agree yes, that with language modernization the [[nodiscard]] is mandatory with this pattern -- it won't be difficult to change everything in eckit::geo, but to change it in the stacks is a whole other problem...